### PR TITLE
80 dockerization

### LINKED ISSUE #80: Build a Docker Container
## Describe the Feature
We're dealing with a lot of environment issues, and new contributors don't have an easy way to set up the app. Between yarn, webpack, and postgres, there are quite a few dependencies that need management.

## Describe the Technical Implementation of the Solution
**Solution**
Create a docker file to automatically build the app. Anyone with docker and git can set up the app easily.

**References**
https://docs.docker.com/compose/rails/

### PR DESCRIPTION
**Pull request by bipster**
This would resolve #80 

This builds on top of https://github.com/ChaelCodes/HuntersKeepers/pull/88

### Dockerfile
This creates an image with node/yarn installed, and a snapshot of the application directory.

### docker-compose.yml
_Two named volumes:_
* data - to hold the Postgres database files
* node - to hold the node_modules/ files

_Four services:_
* db - the Postgres database server, working out of the `data` volume
* node_modules - this builds node_modules/, if necessary, storing the results in the `node_modules` volume.  This is a transient service -- once it's done, it exits out.
* webpacker - the webpack server (depends on `node_modules`)
* web - the rails server (depends on all of the above)

### How the volume mounts work:
**Image After Dockerfile**
* `/hunterskeepers/` contains a copy of the host directory
* `/hunterskeepers/node_modules/` is empty because of the .dockerignore

**node_modules container on initial startup**
* `/hunterskeepers/` contains a copy of the host directory
* `/hunterskeepers/node_modules/` is empty, but bound to the named module `node`

**node_modules container on shutdown**
* `/hunterskeepers/` contains a copy of the host directory
* `/hunterskeepers/node_modules/` is populated, and thus so is the named module `node`

**node_modules container on subsequent startup**
* `/hunterskeepers/` contains a copy of the host directory
* `/hunterskeepers/node_modules/` is non-empty due to the named module `node` having been populated

**web container on startup**
* `/hunterskeepers/` contains a copy of the host directory
* `/hunterskeepers/` gets replaced -- is it now bound to the _actual_ host directory (via the `- './:/hunterskeepers/'` line in the docker-compose)
* `/hunterskeepers/node_modules/` is populated, but with the node packages for the host environment
* `/hunterskeepers/node_modules/` gets replaced -- it is now bound to the cached files the `node_modules` service built into the named `node` volume (via the `- 'node:/hunterskeepers/node_modules/:nocopy'` line in the docker-compose)

Note: Had that last line not had `:nocopy`, the web container's pre-existing `node_modules` directory would be copied into the named `node` volume.  i.e. the changes would go the wrong way.

### Usage
`docker-compose up` should do it ..

For follow-up: document commands to use when needing to rebuild the node dependencies or the database.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# Loads a snapshot of the application files into the image.
+WORKDIR /hunterskeepers
+COPY Gemfile* ./
+RUN gem install bundler:2.1.4 && bundle install
+COPY . /hunterskeepers
+
+# Re-building /node-modules
+RUN rm -fr node_modules \
+ && yarn install --check-files
```
[the remark this answers is not in the document]
Comment (bipster) [reply Dockerfile:23]: Thanks, @exegete46!  Implemented all of the above.  You implicitly clarified a few things I'd been unsure about -- much appreciated.

**Comment by bipster** [comment]
@ChaelCodes Branch updated:
* Static assets should be loading now.  [This line](https://github.com/ChaelCodes/HuntersKeepers/pull/157/files#diff-4e5e90c6228fd48698d074241c2ba760R57) was the issue -- it was set to `0.0.0.0` before, which meant the Rails server static-asset URLs were pointing at :3000 instead of the Webpacker server's :3035.  Changing from `0.0.0.0` to `webpacker` means Docker's internal network will be properly utilized
* Tightened things based on @exegete46 's feedback
* I stopped exposing the Postgres container's server to the host environment, so there won't be the potential awkwardness of two servers fighting over the standard port 5432.

Some timesaver commands:

**To start with a clean slate in docker:**
```
> docker system prune -a
> docker volume rm hunterskeepers_data
> docker volume rm hunterskeepers_node
```
Warning: that first line clears out *all* images/containers, not just the ones related to HK

**To open a Rails CLI:**
```
> docker exec -it hunterskeepers_web_1 bin/rails c
```

**To open a Postgres CLI:**
```
> docker exec -it hunterskeepers_db_1 psql -U postgres hunterskeepers_development
```

**Comment by ChaelCodes** [comment]
Please swap[ filewatcher in development.rb](https://github.com/ChaelCodes/HuntersKeepers/blob/2593267a906c9d064ea614061ea87815b03e7cec/config/environments/development.rb#L67) to ActiveSupport::FileUpdateChecker.

**Comment by bipster** [comment]
> Please swap[ filewatcher in development.rb](https://github.com/ChaelCodes/HuntersKeepers/blob/2593267a906c9d064ea614061ea87815b03e7cec/config/environments/development.rb#L67) to ActiveSupport::FileUpdateChecker.

Done, and drafted documentation explaining that as well as usage notes in the top-level README.md.  Verified the non-evented file-watcher works fine via Docker for Mac.

**Review by ChaelCodes — APPROVED** [review]
✨ This looks GREAT! ✨ I'm really glad that we made this repo significantly more accessible to new contributors! I like the readme docs, and I'll definitely be using this going forward!

**Pull request merged**